apps/aehttp/src/aehttp_dispatch_ext.erl

```diff
@@ -166,6 +166,248 @@ handle_request('GetGenerationByHeight', Params, _Context) ->
+                     #{pubkey => aec_base58c:encode(account_pubkey, Pubkey),
+                       balance => aec_accounts:balance(Account),
+                       nonce => aec_accounts:nonce(Account)}};
+                {error, _} ->
+                    {404, [], #{reason => <<"Account not found">>}}
+            end;
+        {error, _} ->
+            {400, [], #{reason => <<"Invalid public key">>}}
+    end;
+
+handle_request('GetPendingAccountTransactionsByPubkey', Params, _Context) ->
+    case aec_base58c:safe_decode(account_pubkey, maps:get(pubkey, Params)) of
+        {ok, Pubkey} ->
+            case aec_chain:get_account(Pubkey) of
+                {value, _} ->
+                    {ok, Txs0} = aec_tx_pool:peek(infinity, Pubkey),
+                    Txs = [aetx_sign:serialize_for_client_pending(json, T) || T <- Txs0],
+                    JsonTxs = #{data_schema => <<"JSONTx">>,
+                                transactions => Txs},
+                    {200, [], JsonTxs};
+                _ ->
+                    {404, [], #{reason => <<"Account not found">>}}
+            end;
+        {error, _} ->
+            {400, [], #{reason => <<"Invalid public key">>}}
+    end;
+
+handle_request('GetTransactionByHash', Params, _Config) ->
+    case aec_base58c:safe_decode(tx_hash, maps:get(hash, Params)) of
+        {ok, Hash} ->
+            case aec_chain:find_tx_with_location(Hash) of
+                none ->
+                    {404, [], #{<<"reason">> => <<"Transaction not found">>}};
+                {mempool, Tx} ->
+                    JSONTx = aetx_sign:serialize_for_client_pending(json, Tx),
+                    {200, [], JSONTx};
+                {BlockHash, Tx} ->
+                    {ok, Header} = aec_chain:get_header(BlockHash),
+                    JSONTx = aetx_sign:serialize_for_client(json, Header, Tx),
+                    {200, [], JSONTx}
+            end;
+        {error, _} ->
+            {400, [], #{reason => <<"Invalid hash">>}}
+    end;
+
+handle_request('GetTransactionInfoByHash', Params, _Config) ->
+    ParseFuns = [read_required_params([hash]),
+                 base58_decode([{hash, tx_hash, tx_hash}]),
+                 get_transaction(tx_hash, tx),
+                 get_contract_call_object_from_tx(tx, contract_call),
+                 ok_response(
+                    fun(#{contract_call := Call}) ->
+                            aect_call:serialize_for_client(Call)
+                    end)
+                ],
+    process_request(ParseFuns, Params);
+
+handle_request('PostTransaction', #{'Tx' := Tx}, _Context) ->
+    case aehttp_api_parser:decode(tx, maps:get(<<"tx">>, Tx)) of
+        {error, #{<<"tx">> := broken_tx}} ->
+            {400, [], #{reason => <<"Invalid tx">>}};
+        {error, _} ->
+            {400, [], #{reason => <<"Invalid base58Check encoding">>}};
+        {ok, SignedTx} ->
+            %% TODO: lager debug log?
+            aec_tx_pool:push(SignedTx),
```

Comment:
I see that `SignedTx` is the return value of `aetx_sign:deserialize_from_binary` hence is a valid `#signed_tx{}` hence `aec_tx_pool:push/1` can return `ok | {error, _}` (as `pool_db_put`). Not matching on the return value here is misleading for user. Imagine user submitting tx with e.g. too low nonce - what user would see is 200 returned with tx hash and waiting for tx hash to be somewhere in node... though not there. Would it be possible to have at least a minimal check of `aec_tx_pool:push`'s return value in this PR (and if only minimal, tracking rest in distinct Pivotal ticket)?

(The return type of the spec of `aec_tx_pool:push/0` is `ok` though that is incorrect.)

Reply:
https://www.pivotaltracker.com/story/show/159394602